doc/design/python_api.md

```diff
@@ -0,0 +1,186 @@
+
+The block is hierarchical because PaddlePaddle supports RNN and IfElse. For example, RNN is like `for-loop` in programming languages. There is new `block` inside a `for-loop`. To represent hierarchies, `Block` stores the index of `parent Block` inside. The 'index' means the block's position in `Program`'s `blocks`. If `parent_idx=None`, the block itself is the outermost block, i.e., the 'global block'.
+
+
+```python
+class Block(objects):
+    def __init__(self, parent_idx, idx):
+        self.vars = map<string, Variable>()
+        self.ops = vector<Operator>()
+        self.idx = idx
+        self.parent_idx = parent_idx
+    
+    def create_var(self, ...):
+        # create variable in `self.vars`
+        return Variable(...)
+    
+    
+    def create_global_var(self, ...):
+        if self.parent_idx is not None:
+            parent_block = program.get_block(parent_idx)
+            return parent_block.create_global_var(...)
+        else:
+            return self.create_var(...)
+    
+    def create_parameter(self, ...):
+        return self.create_global_var(...)
+    
+    def append_operator(self, ...):
+        self.ops.append(...)
+        
+    def prepend_operator(self, ...):
+       self.ops.prepend(...)
+```
+
+Users are able to create a global variable inside any block since they many create parameters inside a RNN or IfElse. All parameters should be stored in the global block, not the step block in RNN.
+
+Users can create local variables for outputs of operators. Users can also append and prepend an operator in current block. Prepending `random initialize` operator or `load` operator is very useful to initialize parameters before training.
+
+
+### Operator
+
+Operator class will take inputs, outputs and attributes of the operator into `protobuf` OpDesc and create a C++ `OpDesc` instance. The `infer_shape` perform on C++ objects.
+
+```python
+class Operator(object):
+    def __init__(self, type, inputs, outputs, attrs):
+        # create OpDesc in Python
+        op_desc = ...
+        self.cpp_op_desc_ptr = core.OpDesc(op_desc)
+        cpp.infer_shape(self.cpp_op_desc_ptr, inputs, outputs)
+
+    def type(self):
+        return self.cpp_op_desc_ptr.type()
+```
+
+After creating a C++ `OpDesc`, `Operator` in Python can only reads the attribute from C++ side.
+
+### Variable
+
+Operators' inputs, outputs, and parameters are all variables. In our design, a variable has four key attributes: its name(`name`), the block it belongs to(`block`), a pointer pointed to its C++ Protobuf object(`cpp_var_desc_ptr`), and the operator it is created by(`op`). All of these attributes are initialized in the constructor, except the `op`. The `op` will keep being `None` till the variable is taken as an operator's output.
```

Comment:
> The `op` will keep being `None` till the variable is taken as an operator's output.

Do you mean: "The value of `op` is `None` unless the variable is an operator's output."?